In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create Path
path = r'C:\Users\Lenovo\Desktop\Career Foundry\28-06-2024 Instacart Basket Analysis - Saswata Biswas'

In [5]:
# Importing the latest ords_prods_merge data:

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [7]:
#using the loc function on the entire dataset:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [11]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [13]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [17]:
#testing the for loop:

for x in range(30,45):
    print ("My age is %d" %(x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


In [19]:
#checking the frequency or orders_days_of_week:
df_ords_prods_merge['orders_days_of_week'].value_counts(dropna = False)

orders_days_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [25]:
#Creating a loop for seeing busiest days:
result = []

for value in df_ords_prods_merge["orders_days_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [27]:
#In your project brief, you can see that the value 0 means Saturday. This value has the highest frequency, meaning Saturday is the busiest day. Meanwhile, the 4 value has the lowest frequency. A value of 4, here, refers to Wednesday, meaning Wednesday is the slowest day for Instacart app orders.

In [29]:
#Not checking the result as it will be a long list

In [31]:
# Let us create a new Column
df_ords_prods_merge['busiest_day'] = result

In [33]:
#checking to see if the days match with order_days_of_week:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

#### Checked 1, 2, 3, 5 and 6 sum up to 22416875

## Question 02

In [39]:
# Create column “busiest_days" from "orders_day_of_week" to identify two busiest days (0, 1), slowest days (4, 3), or regularly busy (all other days).

result_2 = []

for value in df_ords_prods_merge["orders_days_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [43]:
# Making this into a new Column:
df_ords_prods_merge['busiest_days'] = result_2

In [45]:
#checking to see if the days match with order_days_of_week:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

### Question 04. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [48]:
# Check value counts in "order_hour_of_day" column.

df_ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

The value counts listed above will be split into equal thirds for the following labels: "Most orders" = 9, 10, 11, 12, 13, 14, 15, 16; "Fewest orders" = 23, 0, 1, 2, 3, 4, 5 and 6; "Average orders" = all remaining values in order_hour_of_day.

In [53]:
#Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [55]:
# Create new column "busiest_period_of_day" in ords_prods_merge.

df_ords_prods_merge['busiest_period_of_day'] = result_3

### Print the frequency for this new Column

In [60]:
# Print value counts in "busiest_period_of_day" column.

df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

#### Check count for fewest orders against the hours:
It amounts to 1289137 which is the exact number of the fewest order.

#### Export the Dataframe

In [64]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new.pkl'))